### DISCUSSION TITLE
Configuring an IDE to create scripts

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#589] [Help] [answered]
I started following the tutorial to configure an IDE. I will ask for help here about the points that are not clear to me, and then I will clarify those points in the documentation.

The first point that is not clear to me: at [this section](https://docs.freeplane.org/#/scripting/Scripting_environment_setup?id=install-freeplane_src), it's written "Unpacking may need two steps:", and then the first step is:

> 1. In a program like 7zip open the archive freeplane_src-1.2.13.tar.gz. Most unpackers will allow directly extract into a directory, but jzip creates an intermediate freeplane_src-1.2.13.tar file

My question is if the second step and the gradle part are only necessary if the user uses jzip, and, thus, has a tar file. I used winrar, so I simply extracted freeplane-1.10.4-pre11 to the workspace. How should I proceed?

Thanks,

**Comment by macmarrum** [comment, marked as answer]
I had a look at the page and noticed some content misinterpreted (differences between MediaWiki and Markdown). I fixed it.

I believe that for simple scripting (without debugging on a running Freeplane), the instructions can be simplified. I have something in mind; I'll need to verify it makes sense before posting here.

**Reply by macmarrum** [reply]
I have verified the following _simplified_ procedure for setting up IntelliJ IDEA (this is what I know and use).

### Installation and configuration
1. Go to https://www.jetbrains.com/idea/download/
2. Choose **Community** (direct download link for Windows: https://www.jetbrains.com/idea/download/download-thanks.html?platform=windows&code=IIC)
3. Install IntelliJ IDEA (`Next`, `Next`, `Next`; the default options are fine)
4. Launch IntelliJ IDEA
5. On the Welcome screen, choose `New Project`
6. `Name:` **freeplane-scripting**
7. `Language:` **Java**
8. `Build system:` **IntelliJ**
9. `JDK:` if none detected, select `Download JDK` and choose version **17**. Any `Vendor:` will do. I choose **Eclipse Temurin ...**
1. Click `Create`
1. Once created, go to `File->Project structure (Ctrl+Alt+Shift+S)`
1. Then to `Modules`, and in the right pane to `Dependencies`
1. Click `+` in the `Dependencies` pane, select `1 JARs or Directories`
1. Navigate to your Freeplane installation directory and select `freeplanelauncher.jar` (e.g. `C:\Program Files\Freeplane\freeplanelauncher.jar`)
1. Again `+` `1 JARs or Directories` `C:\Program Files\Freeplane\core\org.freeplane.core\lib`
1. Once more `+` `1 JARs or Directories` `C:\Program Files\Freeplane\plugins\org.freeplane.plugin.script\lib`
2. Close the Project Structure window with `OK` at the bottom.

The installation is done. Now off to creating your first script.

### Writing scripts
In IntelliJ IDEA, right-click on the (blue) folder **src** `New->File` and in the dialog write the name **hello.groovy**.

* In the file (right pane), start writing `def c = ScriptUtils`. A drop-down will appear. 
* Choose the (first) entry pointing to `org.freeplane.plugin.script.proxy`. It will add a line at the top `import org.freeplane.plugin.script.proxy.ScriptUtils`.
* Continue writing in line 3. Put a dot after `ScriptUtils` and wait for a drop-down to appear. Select `c()`.
* In the next line (4), write `def node = ScriptUtils.node()`
* In a new line type `c.stat` and select `statusInfo`. You can use `DOWN` `ENTER` or your mouse.
* Then type ` = "Hello. Current node id: ${node.id}"

#### Running your new script in Freeplane
* Copy your script – select the file in the left pane and `Ctrl+C`
* Go to your `<user-dir>/scripts` and paste it `Ctrl+V`
* Restart Freeplane and run the script to verify it works OK

**Reply by euu2021** [reply]
Thanks, I'm trying to follow your steps.

The script is giving this error:
![image](https://user-images.githubusercontent.com/77707706/186790252-836731c3-e197-42dc-96dd-0e5688970327.png)

In, IDEA, it looks like this:
![image](https://user-images.githubusercontent.com/77707706/186790306-ca97a8b2-616b-4b32-b5c0-3f2c09e29d33.png)

**Reply by macmarrum** [reply]
That's great. You got IntelliJ IDEA working :+1:

It's just a simple typo:
`ScripUtils` :x: 
`ScriptUtils` :heavy_check_mark: 

You'll soon learn that IntelliJ IDEA indicates such errors by a dotted underline.
Also, there is no completion (no drop-down after `.`), when a typo occurs.

**Comment by euu2021** [comment]
> It's just a simple typo

This shows that using an IDE will be useful to me 🤣

---
Additional step that I've done: I pointed the src folder of the IDEA project as an additional folder for FP to look for scripts, so I can skip the part of copying and pasting into the other folder.
